### DISCUSSION TITLE
Relational data gets lost, when the related db is edited – but only when edited for the first time?

### DISCUSSION THREAD

**Discussion by rowild** [directus/directus#13991] [Archive]
Hi!

I use Directus 9 since about 1 month (coming from TYPO3 and strapi) and am very fascinated with this product! awesome interface, easy and very powerful.

But I do observe data loss every once in a while. So far I could not pin down what exactly causes the loss. The data loss I talk about concerns data in m:n relations in a library system. 

In my case, a "piece of art", called "opus" (be it a book, painting, composition, etc) has many m:n relations to e.g. categories, events, related_people, variants etc.

It seems that as soon as I edit a related category FOR THE FIRST TIME only, all related data is emptied in "opus". But it is not in the related table as the screenshot demonstrates, which has empty relations. (The other relations that can be seen, have been added again.) 
However, if I edit a relational table again (e.g. adding another field to the opus_category table, maybe deleting another one), all the relational data stay alive.

![Bildschirmfoto 2022-06-20 um 10 51 17](https://user-images.githubusercontent.com/213803/174565396-853d006f-2034-4282-875d-aa40d4479d35.png)

As I said above: I have yet no real clue of how to reliably reproduce this scenario. Therefore this topic is a bit vague ATM. But I ask to keep it open - maybe other people who have similar experiences, can add their observations and scenarios. I, of course, intent to update as soon as I find out more.

Whatever it is, it must be found and solved. Because losing data is, as you most likely agree, a no-go.

PS: I have a "gut feeling" that this might have to do with the "Required value to be set on creation" (which is a weird option, since it only checks the very first time; why not always? Because, when I edit such data, I can do whatever I want – leave it empty, enter fantasy text... no check anymore after the first time...)

SOFTWARE I use:

Directus 9.12.2
SQLite 3.28.0
Node 16.14.2
MacOS 10.15.6 Catalina

**Comment by azrikahar** [comment]
Just inquiring some more info as I believe they may be relevant:

- What Directus version is this?

- What database vendor are you using in this setup? SQLite or something else?

- This isn't compulsory, and probably will not be as relevant once you do update or find out more, but would you mind providing a quick video capture of the workflow mentioned, as well as showing the raw value (click on the "Opus" label and you should see `Edit Raw Value`, and do scroll the content if possible)?

> "Required value to be set on creation" (which is a weird option, since it only checks the very first time; why not always? Because, when I edit such data, I can do whatever I want – leave it empty, enter fantasy text... no check anymore after the first time...)

Totally understandable! We are currently tracking this in #11057.

**Reply by rowild** [reply]
@azrikahar Thank you for your quick feedback! I update the infos about software versions in the main post. The video will come soon (working on it)!

**Reply by azrikahar** [reply]
Thanks for updating with the detailed info! I'm now suspecting it's related to #13936 which is specific to SQLite only (it was technically fixed, but could be a regression from dependencies). 

You did mention:

> However, if I edit a relational table again (e.g. adding another field to the opus_category table, maybe deleting another one), all the relational data stay alive.

But by "stay alive", do you mean they are still present, but the junction table rows are turned to null?

**Reply by rowild** [reply]
@azrikahar Here is a video:
https://dl.rowild.at/directus-9-api-m2m-delete-problem.mp4
Sorry, but after 5 tries I still won't get it below 10MB. And of course I am just not a video guy, my English is bad and the whole thing is still confuse. Nevertheless I hope it helps a bit, since I think it pinpoints the problem.

Here a little summary of what I could find out:
- The problem arises, when a field gets DELETED on either side of the m2m relational tables (it does NOT happen, when a table is added)
- The problem happens ALWAYS, when a field is deleted
- After deleting a field, the [table]_id of the table that was just edited, is set to NULL in the junction table

It pretty much seems to fit with what is described in the issue you mention above.
Thank you!

PS: I will now try the same setup with postgres and mysql and report back ASAP.

**Reply by rowild** [reply]
I just tried postgres 14 and there this error does NOT happen! 
I will now try mysql / mariadb and report back ASAP.

**Reply by azrikahar** [reply]
@rowild no worries, you went above and beyond with added commentaries (and I can understand you very clearly), really appreciate the effort! 

Narrowing it down to only when a field gets _deleted_ does pretty much confirms that it's a SQLite-specific issue, since what's happening is SQLite will technically drop the table and re-create it _without_ said column. Previously this was fixed but unfortunately it's now popping up again recently. Feel free to subscribe to #13936 for future updates on this issue 👍

**Reply by rowild** [reply]
@azrikahar Excellent! I am already subscribed to that issue, thank you!
(I finished the mysql test before reading this, and everything is working fine there, as you mentioned. :-) )
